File: Java-Core/src/main/java/presto/declaration/NativeMethodDeclaration.java
package presto.declaration;

import presto.error.PrestoError;
import presto.grammar.ArgumentList;
import presto.grammar.Identifier;
import presto.runtime.Context;
import presto.statement.IStatement;
import presto.statement.StatementList;
import presto.type.IType;
import presto.type.VoidType;
import presto.utils.CodeWriter;
import presto.value.IValue;

public class NativeMethodDeclaration extends ConcreteMethodDeclaration {

	public NativeMethodDeclaration(Identifier name, ArgumentList arguments, IType returnType, StatementList instructions) {
		super(name,arguments,returnType, instructions);
	}

	@Override
	public IValue interpret(Context context) throws PrestoError {
		context.enterMethod(this);
		try {
			return statements.interpret(context, true);
		} finally {
			context.leaveMethod(this);
		}
	}
	
	@Override
	protected void toPDialect(CodeWriter writer) {
		writer.append("def native ");
		writer.append(name);
		writer.append(" (");
		arguments.toDialect(writer);
		writer.append(")");
		if(returnType!=null && returnType!=VoidType.instance()) {
			writer.append("->");
			returnType.toDialect(writer);
		}
		writer.append(":\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
	}

	@Override
	protected void toODialect(CodeWriter writer) {
		if(returnType!=null  && returnType!=VoidType.instance()) {
			returnType.toDialect(writer);
			writer.append(" ");
		}
		writer.append("native method ");
		writer.append(name);
		writer.append(" (");
		arguments.toDialect(writer);
		writer.append(") {\n");
		writer.indent();
		for(IStatement statement : statements) {
			statement.toDialect(writer);
			writer.newLine();
		}
		writer.dedent();
		writer.append("}\n");
	}
	
	@Override
	protected void toEDialect(CodeWriter writer) {
		writer.append("define ");
		writer.append(name);
		writer.append(" as: native method ");
		arguments.toDialect(writer);
		if(returnType!=null && returnType!=VoidType.instance()) {
			writer.append("returning: ");
			returnType.toDialect(writer);
			writer.append(" ");
		}
		writer.append("doing:\n");
		writer.indent();
		statements.toDialect(writer);
		writer.dedent();
		writer.append("\n");
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/declaration/NativeMethodDeclaration.js
var ConcreteMethodDeclaration = require("./ConcreteMethodDeclaration").ConcreteMethodDeclaration;
var VoidType = require("../type/VoidType").VoidType;

function NativeMethodDeclaration(name, args, returnType, stmts) {
	ConcreteMethodDeclaration.call(this, name, args,returnType, stmts);
	return this;
}

NativeMethodDeclaration.prototype = Object.create(ConcreteMethodDeclaration.prototype);
NativeMethodDeclaration.prototype.constructor = NativeMethodDeclaration;

NativeMethodDeclaration.prototype.check = function(context) {
	var checked = this.fullCheck(context, true);
	return this.returnType != null ? this.returnType : checked;
};

NativeMethodDeclaration.prototype.interpret = function(context) {
	context.enterMethod(this);
	try {
		var result = this.statements.interpretNative(context, this.returnType);
        return this.castToReturnType(context, result);
	} finally {
		context.leaveMethod(this);
	}
};

NativeMethodDeclaration.prototype.castToReturnType = function(context, value) {
    // can only cast to specified type, and if required
    if(this.returnType!=null && !(value.type.isAssignableTo(context, this.returnType))) {
        // only cast if implemented, on a per type basis
        if(this.returnType.nativeCast)
            value = this.returnType.nativeCast(context, value);
    }
    return value;
};

NativeMethodDeclaration.prototype.toSDialect = function(writer) {
    writer.append("def native ");
    writer.append(this.name);
    writer.append(" (");
    this.args.toDialect(writer);
    writer.append(")");
    if(this.returnType!=null && this.returnType!=VoidType.instance) {
        writer.append("->");
        this.returnType.toDialect(writer);
    }
    writer.append(":\n");
    writer.indent();
    this.statements.toDialect(writer);
    writer.dedent();
};

NativeMethodDeclaration.prototype.toODialect = function(writer) {
    if(this.returnType!=null  && this.returnType!=VoidType.instance) {
        this.returnType.toDialect(writer);
        writer.append(" ");
    }
    writer.append("native method ");
    writer.append(this.name);
    writer.append(" (");
    this.args.toDialect(writer);
    writer.append(") {\n");
    writer.indent();
    for(var i=0; i<this.statements.length; i++) {
        this.statements[i].toDialect(writer);
        writer.newLine();
    }
    writer.dedent();
    writer.append("}\n");
};

NativeMethodDeclaration.prototype.toEDialect = function(writer) {
    writer.append("define ");
    writer.append(this.name);
    writer.append(" as: native method ");
    this.args.toDialect(writer);
    if(this.returnType!=null && this.returnType!=VoidType.instance) {
        writer.append("returning: ");
        this.returnType.toDialect(writer);
        writer.append(" ");
    }
    writer.append("doing:\n");
    writer.indent();
    this.statements.toDialect(writer);
    writer.dedent();
    writer.append("\n");
};

exports.NativeMethodDeclaration = NativeMethodDeclaration;
